### PR TITLE
[main][feat] change liquidation prize logic

### PR DESCRIPTION
**Pull request by abc-alpaca**
## Description
 Change Liquidator prize 

## Why?
 Adjust smart contract for liquidation strategy

### CODE REVIEW

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 1]
```diff
@@ -65,27 +71,43 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
...
+    );
   }
 
   /// @dev Set all the basic parameters. Must only be called by the owner.
   /// @param _minDebtSize The new minimum debt size value.
   /// @param _reservePoolBps The new interests allocated to the reserve pool value.
   /// @param _killBps The new reward for killing a position value.
   /// @param _interestModel The new interest rate model contract.
+  /// @param _buyBackBps The reward for buyback and burn after successfully killing a position.
+  /// @param _buyBack The reward for buybackAnd Burn Address
```
Comment (almost-llamas) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:100]: `the address where buy back and burn portion transferred to`

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 2]
```diff
@@ -65,27 +71,43 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
...
+      _buyBack
+    );
   }
 
   /// @dev Set all the basic parameters. Must only be called by the owner.
   /// @param _minDebtSize The new minimum debt size value.
   /// @param _reservePoolBps The new interests allocated to the reserve pool value.
   /// @param _killBps The new reward for killing a position value.
   /// @param _interestModel The new interest rate model contract.
+  /// @param _buyBackBps The reward for buyback and burn after successfully killing a position.
```
Comment (almost-llamas) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:99]: `The portion of reward for buy back and burn after successfully killing a position.`

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 3]
```diff
@@ -57,6 +59,10 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   uint256 public maxKillBps;
   /// list of whitelisted callers
   mapping(address => bool) public override whitelistedCallers;
+  // The reward for buyback and burn after successfully killing a position.
+  uint256 public override getBuyBackBps;
+  // The address of buyback
```
Comment (almost-llamas) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:64]: `The address where buyback and burn portion will be transferred to.`

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 4]
```diff
@@ -57,6 +59,10 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   uint256 public maxKillBps;
   /// list of whitelisted callers
   mapping(address => bool) public override whitelistedCallers;
+  // The reward for buyback and burn after successfully killing a position.
```
Comment (almost-llamas) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:62]: `The portion of reward for buyback and burn after successfully killing a position.`

**Review by almost-llamas — CHANGES_REQUESTED** [review]
Please also adjust the deployment script as the constructor of the contract has been changed.

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 5]
```diff
@@ -57,6 +59,10 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   uint256 public maxKillBps;
   /// list of whitelisted callers
   mapping(address => bool) public override whitelistedCallers;
+  // The portion of reward for buyback and burn after successfully killing a position.
+  uint256 public override getBuybackBps;
```
Comment (spicysquid168) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:63]: `getKillTreasuryBps` would be better reflect the real meaning as we will use this part of fee to reserve for bad debt. Then if no bad debt happened, will be used for buyback later.

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 6]
```diff
@@ -57,6 +59,10 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
...
+  // The address where buyback and burn portion will be transferred to.
+  address public buyback;
```
Comment (spicysquid168) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:65]: Please use `treasuryAccount` instead

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 7]
```diff
@@ -171,4 +197,9 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   function killFactor(address worker, uint256 debt) external view override returns (uint256) {
     return workers[worker].killFactor(worker, debt);
   }
+
+  /// @dev return the buyback Address
+  function getBuybackAddr() external view override returns (address) {
+    return buyback;
```
Comment (spicysquid168) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:203]: Please also handle the case where `treasuryAccount` haven't been set yet (if `treasuryAccount == address(0)`), then return Deployer address first (0xC44f82b07Ab3E691F826951a6E335E1bC1bB0B51)

Comment (spicysquid168) [reply contracts/6/protocol/ConfigurableInterestVaultConfig.sol:203]: This case can happen as we will upgrade the implementation contract of VaultConfig, then after upgrade the `treasuryAccount` will has `address(0)` value.

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 8]
```diff
@@ -65,27 +71,43 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
...
     uint256 _killBps,
     InterestModel _interestModel,
     address _wrappedNative,
     address _wNativeRelayer,
-    address _fairLaunch
+    address _fairLaunch,
+    uint256 _buybackBps,
+    address _buyback
   ) public onlyOwner {
     require(_killBps <= maxKillBps, "ConfigurableInterestVaultConfig::setParams:: kill bps exceeded max kill bps");
```
Comment (spicysquid168) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:112]: This require should be updated to `require(_killBps.add(_getKillTresauryBps) <= maxKillBps, "...")` as `_getKillTreasuryBps` is a part of total `killBps`

File: contracts/6/protocol/SimpleVaultConfig.sol [hunk 9]
```diff
@@ -45,6 +45,10 @@ contract SimpleVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   address public fairLaunch;
   /// list of whitelisted callers
   mapping(address => bool) public override whitelistedCallers;
+  // The reward for buyback and burn after successfully killing a position.
+  uint256 public override getBuybackBps;
```
Comment (spicysquid168) [inline contracts/6/protocol/SimpleVaultConfig.sol:49]: please apply the same comments as `ConfigurableInterestVaultConfig` above

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

File: contracts/6/protocol/Vault.sol [hunk 10]
```diff
@@ -347,18 +347,32 @@ contract Vault is IVault, ERC20UpgradeSafe, ReentrancyGuardUpgradeSafe, OwnableU
     uint256 beforeToken = SafeToken.myBalance(token);
     IWorker(pos.worker).liquidate(id);
     uint256 back = SafeToken.myBalance(token).sub(beforeToken);
-    uint256 prize = back.mul(config.getKillBps()).div(10000);
+
+    uint256 liquidatorPrize = back.mul(config.getKillBps()).div(10000);
+    uint256 buybackBurnPrize = back.mul(config.getBuybackBps()).div(10000);
```
Comment (spicysquid168) [inline contracts/6/protocol/Vault.sol:352]: should rename to `tresauryFees`

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

File: test/CakeMaxiWorker.test.ts [hunk 11]
```diff
@@ -69,6 +69,8 @@ describe('CakeMaxiWorker', () => {
   const poolId = 0
   const WORK_FACTOR = '7000';
   const KILL_FACTOR = '8000';
+  const BUYBACK_BPS = '100'; // 1% Buyback & Burn (success)
+  const BUYBACK_OVERFLOW_BPS = '500' //5% Buyback & Burn (10% prize + 5% BB causes bad debt)
```
Comment (spicysquid168) [inline test/CakeMaxiWorker.test.ts:73]: Better simulate by swap a huge FTOKEN -> BTOKEN than force BUYBACK_OVERFLOW

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol (hunk 8 above)
Comment (abc-alpaca) [reply contracts/6/protocol/ConfigurableInterestVaultConfig.sol:112]: I using + instead of add

File: contracts/6/protocol/Vault.sol [hunk 12]
```diff
@@ -219,8 +219,8 @@ contract Vault is IVault, ERC20UpgradeSafe, ReentrancyGuardUpgradeSafe, OwnableU
     uint256 amount = share.mul(totalToken()).div(totalSupply());
     _burn(msg.sender, share);
     if (token == config.getWrappedNativeAddr()) {
-      SafeToken.safeTransfer(token, config.getWNativeRelayer(), amount);
-      WNativeRelayer(uint160(config.getWNativeRelayer())).withdraw(amount);
+      SafeToken.safeTransfer(token, config.getWNativeRelayerAddr(), amount);
```
Comment (spicysquid168) [inline contracts/6/protocol/Vault.sol:222]: please keep the above lines the same. `config.getWNativeRelayer()`

File: contracts/6/protocol/interfaces/IVaultConfig.sol [hunk 13]
```diff
@@ -24,7 +24,7 @@ interface IVaultConfig {
   function getWrappedNativeAddr() external view returns (address);
 
   /// @dev Return the address of wNative relayer.
-  function getWNativeRelayer() external view returns (address);
+  function getWNativeRelayerAddr() external view returns (address);
```
Comment (spicysquid168) [inline contracts/6/protocol/interfaces/IVaultConfig.sol:27]: Keep the interface here the same

File: contracts/6/protocol/SimpleVaultConfig.sol [hunk 14]
```diff
@@ -36,24 +36,30 @@ contract SimpleVaultConfig is IVaultConfig, OwnableUpgradeSafe {
...
   uint256 public override getKillBps;
   /// Mapping for worker address to its configuration.
-  mapping (address => WorkerConfig) public workers;
+  mapping(address => WorkerConfig) public workers;
   /// address for wrapped native eg WBNB, WETH
-  address public wrappedNativeAddr;
+  address public override getWrappedNativeAddr;
   /// address for wNative relater
-  address public wNativeRelayer;
+  address public override getWNativeRelayerAddr;
```
Comment (spicysquid168) [inline contracts/6/protocol/SimpleVaultConfig.sol:43]: should be `getWNativeRelayer`

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 15]
```diff
@@ -60,20 +60,20 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   /// list of whitelisted callers
   mapping(address => bool) public override whitelistedCallers;
   // The portion of reward for buyback and burn after successfully killing a position.
-  uint256 public override getBuybackBps;
+  uint256 public override getKillTreasuryBps;
   // The address where buyback and burn portion will be transferred to.
```
Comment (almost-llamas) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:64]:
```
// address of treasury account

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 16]
```diff
@@ -60,20 +60,20 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   /// list of whitelisted callers
   mapping(address => bool) public override whitelistedCallers;
   // The portion of reward for buyback and burn after successfully killing a position.
```
Comment (almost-llamas) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:62]:
```
  // The portion of reward that will be transferred to treasury account  after successfully killing a position.

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 17]
```diff
@@ -96,30 +96,34 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   /// @param _reservePoolBps The new interests allocated to the reserve pool value.
   /// @param _killBps The new reward for killing a position value.
   /// @param _interestModel The new interest rate model contract.
-  /// @param _buybackBps The portion of reward for buyback and burn after successfully killing a position.
-  /// @param _getbuybackAddr The address where buyback and burn portion will be transferred to.
+  /// @param _getKillTreasuryBps The portion of reward for buyback and burn after successfully killing a position.
+  /// @param _treasury The address where buyback and burn portion will be transferred to.
```
Comment (almost-llamas) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:100]: revise the comment, referring to previous comments

File: contracts/6/protocol/SimpleVaultConfig.sol [hunk 18]
```diff
@@ -46,20 +46,20 @@ contract SimpleVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   /// list of whitelisted callers
   mapping(address => bool) public override whitelistedCallers;
   // The reward for buyback and burn after successfully killing a position.
-  uint256 public override getBuybackBps;
+  uint256 public override getKillTreasuryBps;
   // The address of buyback
-  address public override getBuybackAddr;
+  address public treasury;
```
Comment (almost-llamas) [inline contracts/6/protocol/SimpleVaultConfig.sol]: revise the comment, referring to previous comments

File: contracts/6/protocol/interfaces/IVaultConfig.sol [hunk 19]
```diff
@@ -51,8 +51,8 @@ interface IVaultConfig {
   function killFactor(address worker, uint256 debt) external view returns (uint256);
 
   /// @dev Return the Buyback value
-  function getBuybackBps() external view returns (uint256);
+  function getKillTreasuryBps() external view returns (uint256);
 
   /// @dev Return The address where buyback and burn portion will be transferred to.
-  function getBuybackAddr() external view returns (address);
+  function getTreasuryAddr() external view returns (address);
```
Comment (almost-llamas) [inline contracts/6/protocol/interfaces/IVaultConfig.sol]: revise the comment, referring to previous comments

**Review by almost-llamas — DISMISSED** [review]

File: deploy/008_deploy_configurable_interest_vault_config.ts [hunk 20]
```diff
@@ -17,7 +17,9 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
   */
   const MIN_DEBT_SIZE = ethers.utils.parseEther('100');
   const RESERVE_POOL_BPS = '1000';
-  const KILL_PRIZE_BPS = '500';
+  const KILL_PRIZE_BPS = '250';
+  const BUYBACK_BPS = '250';
```
Comment (spicysquid168) [inline deploy/008_deploy_configurable_interest_vault_config.ts:21]: TREASURY_KILL_BPS?

File: deploy/008_deploy_configurable_interest_vault_config.ts [hunk 21]
```diff
@@ -17,7 +17,9 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
...
+  const BUYBACK_ADDR = '';
```
Comment (spicysquid168) [inline deploy/008_deploy_configurable_interest_vault_config.ts:22]: TREASURY_ADDR?

File: contracts/6/protocol/ConfigurableInterestVaultConfig.sol [hunk 22]
```diff
@@ -171,4 +187,9 @@ contract ConfigurableInterestVaultConfig is IVaultConfig, OwnableUpgradeSafe {
   function killFactor(address worker, uint256 debt) external view override returns (uint256) {
     return workers[worker].killFactor(worker, debt);
   }
+
+  /// @dev Return the treasuryAddr
+  function getTreasuryAddr() external view override returns (address) {
+    return treasury == address(0) ? 0xC44f82b07Ab3E691F826951a6E335E1bC1bB0B51 : treasury;
```
Comment (spicysquid168) [inline contracts/6/protocol/ConfigurableInterestVaultConfig.sol:193]: Please add test-case that cover this line too.

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

**Review by almost-llamas — APPROVED** [review]

**Review by spicysquid168 — APPROVED** [review]

**Pull request merged**